Desktop: Merge wp-desktop in to Calypso - First Steps

spen opened the pull request:
Phew! I was snagged for some time, but I _think_ that I've shaken free with this PR!

This PR is to be the first stepping stone towards bringing the existing `wp-desktop` project in to Calypso, ultimately allowing for more solid development & deployment processes.

I've attempted to make this change reasonably minimal and so, other than migrating the `wp-desktop` code, there should be no deviations or added behaviour.

There's still a way to go to achieve the the benefits that we are aiming towards with the merge of the `wp-desktop` project but having the projects merged should give a good entry point for starting to implement better solutions :D

### Motivations	

- Give the desktop client better visibility to Calypso developers.
- Encourage devs to develop & test with the desktop client more (or even as their primary client).
- Ability to include the desktop build & release in to Calypsos release process.
- Help spot and fix desktop specific bugs faster.

### Testing
#### Development Build
`cd` in to `/desktop` and run `make run`
This will take some time, particularly the first time around as both server and client bundles will be generated, along with a number of files that, for now, will be copied in to `desktop/build` from Calypso itself.

Watch the terminal output for errors, or anything that seems unusual.

Finally - once every build process is complete, you should see Calypso automatically launch in an Electron shell.

#### Release Builds
Following the instructions [found here](https://github.com/Automattic/wp-desktop/blob/master/docs/release.md) in `wp-desktop` a release should be successfully built for **mac**, **windows** and **linux**.
Note: I'm still looking in to how to test this myself.

#### Tests
Both `make test` & `make test-osx` should pass.
Currently, `make test` fails and `make test-osx` errors for me while attempting to sign the app.

All other Make commands should also work

### Reaching A Merge-able Milestone

First and foremost, each of the `make` commands should run as they would in `wp-desktop`
- [x] dev
  - [x] `make run`
- [x] production
  - [x] `make osx`
  	- [x] Successfully Builds
  	- [x] Tested
  - [x] `make linux`
  	- [x] Successfully Builds
  	- [x] Tested
  - [x] `make win32`
  	- [x] Successfully Builds
  	- [x] Tested
- [x] tests
	- [x] `make test`
	- [x] `make test-osx` ( #13295 )
- [ ] packaging
	- [x] `make package-osx`
	- [ ] `make package-linux`
	- [x] `make package-win32`


Once the above points are in order all of `wp-desktop`'s functions will be available directly in Calypso and so we could consider landing this branch and deprecating `wp-desktop` - though there are some minor improvements that we might consider merging in to the feature branch first:

- [x] Improve `desktop/webpack.shared.js` resolvers. ( #12537 )
-  ~Merge `desktop/package.json` in to `root/package.json`~ ( #12749 )
	Unfortunately this can't be done, at least not in a simple manner ([notes here](https://github.com/Automattic/wp-calypso/pull/12749#issuecomment-292300949))
- [x] Add convenience `make` commands to root Makefile. ( #12535 )
- [ ] Documentation
  - [x] Update references in existing docs ( #12487 )
  - [x] Update desktop/README.md` to better fit new context. ( #12487 )
  - [ ] Fold `desktop/CREDITS.md` in to the root. ( #13373 )
  - [x] Fold `CONTRIBUTING.md` and `LICENSE.md` files to root. ( #12487 )
  - [x] Document any new behaviour ( #12487 `make desktop-run` only)
- [x] Address the myriad of linting errors ( #13641 )
- [ ] Update/move the [circle.yml](https://github.com/Automattic/wp-desktop/blob/master/circle.yml) file ( #13325 )

Beyond this, there are a number of improvements that we could start to make post-merge - some of the tasks and changes that we should aim to land soon after merging:

- Forward existing [PRs](https://github.com/Automattic/wp-desktop/pulls) and [Issues](https://github.com/Automattic/wp-desktop/issues) from `wp-desktop` to `wp-calypso`.
- Improvements to directory structure & file locations
- Explore ways to trim the process
  - Perhaps we can `cp` less files and instead be more direct
- Reduce or eliminate the reliance on `desktop/webpack.config.js`, instead utilising the configs already found in Calypso.


I'd appreciate any suggests that you might think of while reviewing - even if they seem out of scope for the initial merge. If they are out of scope, I'll make note of them for future reference. 😃

mtias commented on desktop/desktop/lib/assets/index.js:
```diff
@@ -0,0 +1,9 @@
+const path = require( 'path' );
```
`assets` seems like it should belong elsewhere.

mtias commented on desktop/desktop/lib/calypso-commands/README.md:
```diff
@@ -0,0 +1,13 @@
+Calypso Commands
```
This module should probably be called `ipc-commands` not calypso.

mtias commented on desktop/desktop/lib/config/README.md:
```diff
@@ -0,0 +1,14 @@
+Config
```
`config` should probably live in `desktop/config`.

mtias commented on desktop/desktop/lib/cookie-auth/index.js:
```diff
@@ -0,0 +1,137 @@
+function authorize( username, token ) {
+	var responder = new events.EventEmitter();
+	var body = 'log=' + username;
+	var options = url.parse( 'https://wordpress.com/wp-login.php' );
```
There's a `login_url` config setting in the main config file that should be used here.

mtias commented on desktop/desktop/lib/cookie-auth/index.js:
```diff
@@ -0,0 +1,137 @@
+'use strict';
```
Like other `lib` modules, it seems we could move this one to the main `client/lib`.

spen replied:
@mtias I'm looking at tackling the directory layout next and so wanted to ask your opinion on this first - do you mean all `desktop/desktop/lib` modules should be moved to `client/lib`, so we can get rid of the `desktop/desktop/lib` modules?

mtias replied:
I think so, but I haven't reviewed them all. Some things in desktop lib may not really be lib stuff as well.

mtias commented on desktop/desktop/lib/state/README.md:
```diff
@@ -0,0 +1,6 @@
+State
```
This seems like it should be placed at the root of the desktop folder.

spen replied:
Thanks for taking a look at this so quickly @mtias, for changes like this - what do you think a good strategy should be? Making them here, as part of this PR, or in `wp-desktop` before migrating it over?
To me making the changes in `wp-desktop` first seems logical, but we would lose some git history in doing so.
I'm also concerned with making too many changes to the structure while `wp-desktop` is being maintained as any deviations might cause headaches as changes are made there that need to be echoed here.

nb replied:
I agree we need a decision here – @spen, do you think we will be able to quickly enough make everything work in the Calypso repo, so that we merge a working version of this and address the rest of the non-blocker feedback (like @mtias’s many suggestions) in subsequent PRs. What’s the minimum amount of changes needed to move completely from the other repo?

spen replied:
I think a decent strategy might be to continue with these suggestions in PR's based off of this branch (vs master)... That way we can keep PR's well scoped - avoiding a situation where reviewers are constantly reviewing a moving target, but at the same time we're not rushed in to merging to master in a state that's less than ideal.
This is what I've started doing with some of the other changes that I've been working on, like #12537.

My only worry with doing it this way is that work is still being done over at `wp-desktop` and making sure that this work is included in the merge is a fairly manual task... one that will become harder as the two become further out of sync. Luckily the rate of change there is much slower so this might not be a huge issue.

mtias commented on desktop/public_desktop/about.html:
```diff
@@ -0,0 +1,121 @@
+<!DOCTYPE html>
```
Could we use the [main public location](https://github.com/Automattic/wp-calypso/tree/master/public) for these?

nb replied:
👍 We can move so much out of the desktop directory…

mtias commented:
Wonder if the assets should be kept in another repository. Assets like images tend to grow the size of it for no good reason.

spen commented:
> Wonder if the assets should be kept in another repository. Assets like images tend to grow the size of it for no good reason.

I've hit issues like this in past projects, so I'd be keen to do _something_ with assets.
I do vaguely remember seeing that GitHub offer some sort of service that handles assets which could be worth exploring - It was probably a year or two ago when I saw this though and didn't explore it too much.

mtias commented:
What's the ideal flow we expect developers to take to test the app—say, for the current branch they are working on in core Calypso?

spen commented:
From my understanding the _ideal_ is to be able to use the Electron app as their primary dev client if they want to. It would ideally use hot reloading and the likes to be just as smooth a process as running Calypso in the browser (obviously we're still a way off this).

mtias commented:
Does it make more sense to move the make commands to the main make file for the project?

spen commented:
I was planning to make some changes after this PR that would add some convenience commands similar (or even identical) to those that I've used while playing with this task, like so:

```
make desktop-run:
  @cd desktop; make run
```

There is of course the option of merging the whole Makefile from `/desktop` in to Calypso's Makefile but I was worried that this might add a lot of extra noise to a file that might already be daunting to some. That said, so long as the desktop specific commands are kept neatly separate it shouldn't be too heavy to comprehend.

mtias commented:
@spen are you referring to https://github.com/blog/1986-announcing-git-large-file-storage-lfs for storing assets?

spen commented:
That's the one! I'm not sure if it's so suitable here though unfortunatley :/ I've still not dove right in to reading about it but it seems like we have many small assets vs a few large ones right?
I'm wondering what a good alternative would look like for managing the assets... We'd need to be able to pull these assets in during the build processes - if that place that we're pulling from is another repo then it seems that we'd be back to having a git submodule in the set up, something I feel people want to avoid - I'll keep pondering on this one!

spen commented:
Pinging @astralbodiesm, and a few others who I think may be interested... [ @samouri, @gziolo, @mkaz, @johngodley ] 😀

nb commented:
Thanks, @spen, as a start, without any extra hassle, I managed to build the app just by running `cd desktop; make run` 🎈 More specific comments to follow.

astralbodies commented:
I'll definitely give this a whirl with code signing and testing the deployments to each platform on Monday!

nb commented:
How does `make test` fail for you?

spen commented:
Adding 'blocked' status to this PR as I'm planning to update the desktop code with the final changes from https://github.com/Automattic/wp-desktop/pull/275 before merging.

spen commented:
> How does make test fail for you?

Running `make test` actually fails for reasons different to those mentioned in https://github.com/Automattic/wp-desktop/issues/281 - This looks to be another issue with resolving/`NODE_PATH`s - hopefully a quick fix now that I have a better idea of these issues!

spen commented:
> How does make test fail for you?

It turned out that the test was failing due to a template that couldn't be found - This was because the `CALYPSO_PATH` was no longer pointing to a valid location. 
Updating this to just `pwd` fixed the test.

I still need to take a closer look at the issue @astralbodies pointed out... By bypassing the code-signing code I should be able to reproduce it!

spen commented:
> During the conversion most files turned out with their executable bit up.

Well spotted - something that I wasn't even aware could happen and so definitely wouldn't have noticed myself!
In trying to fix this across all files I decided to kill 2 birds with one stone and merge in the latest changes from `wp-desktop`, as this was something I had planned to do next anyway.
I made a commit that only addressed the executable bit of each file and squashed that in to the first commit of this PR, effectively overriding the initial transfer of the `wp-desktop` files, then replayed any commits since on to that commit :)

spen commented:
> I still need to take a closer look at the issue @astralbodies pointed out... By bypassing the code-signing code I should be able to reproduce it!

Unfortunately it doesn't look like I'm able to reproduce by simply skipping `codesign`ing step.
I'll continue to look at other workarounds.

<img width="532" alt="screen shot 2017-03-29 at 17 23 39" src="https://cloud.githubusercontent.com/assets/4335450/24465303/79c90e9e-14a4-11e7-9086-0f9bc60a2c71.png">

ockham commented on desktop/builder.js:
```diff
@@ -0,0 +1,66 @@
+			builder.cleanUp( path.join( __dirname, 'release' ), opts );
+		}
+	} );
+} )
```
A [super-long `Makefile`](https://github.com/Automattic/wp-calypso/pull/12430/files#diff-b1cd36ac3f156ef4dc61bff0f03cf429), 3 webpack configs (in addition to Calypso's own), and a `builder.js`. I really look forward to hopefully dumping some of those...

spen commented:
Interestingly, when I comment out the `require` and usage of `superagent` I see a similar error regarding express.

<img width="532" alt="screen shot 2017-04-05 at 22 34 47" src="https://cloud.githubusercontent.com/assets/4335450/24728071/5b40503e-1a50-11e7-95e6-92d25373eed0.png">

I had a feeling this might be the case and by looking a little closer into the app's contents it looks like node modules should be installed in `(app)/Contents/Resources/app`.
Somewhere along the process I must have missed a step that installs/moves these dependencies in the right place.
Checking that out next 🙂

spen commented:
> Somewhere along the process I must have missed a step that installs/moves these dependencies in the right place.

It turns out that [I removed a step](https://github.com/Automattic/wp-calypso/pull/12430/commits/0d29138e9f0bb5adb7c7b0c456b23e1ecf406364#diff-b1cd36ac3f156ef4dc61bff0f03cf429L106) that's critical in `Makefile:package` - adding that line back in resolves the issue! 😀

Alas, that's not the end of my woes 😩 - I'm now seeing an error where `index.jade` can not be found:

![views-error](https://cloud.githubusercontent.com/assets/4335450/24806190/efc9e79a-1bab-11e7-87cb-3ac5e240a294.png)

Two things are interesting here...

The obvious being that the views are suddenly not found - they're found with `make run` without issue and when inspecting the app contents from this branch against the contents of the app produced by `wp-desktop` the views are in the same place.

Second, looking at [this condition](https://github.com/Automattic/wp-calypso/blob/master/server/render/index.js#L100-L104) we would expect the error message to state that `desktop.jade` could not be found as `desktop` config should be 'enabled'.
I inferred from this that the config files are not being pointed to correctly either and after some investigation found that none of the config files are found... 

Why not? 🤔

After a lot of trial and error I eventually decided to just log out the value of `__dirname` from `server/config/index.js` in both projects - I looked here because this file sets the `configPath` variable for locating the various config files with this line:
```js
const configPath = path.resolve( __dirname, '..', '..', 'config' );
```

I also logged the result of this `path.resolve`, the result of `path.resolve( __dirname )` and `process.cwd()`.

Here's what I found:

|  | wp-desktop | wp-calypso |  
| --- | --- | --- |
| process.cwd() | Contents/Resources/app | Contents/Resources/app |
| __dirname | calypso/server/config | ../server/config 👀 |
| resolve( __dirname ) | Contents/Resources/app/calypso/server/config | Contents/Resources/server/config |
| resolve( __dirname, '..', '..', 'config' ) | Contents/Resources/app/calypso/config | Contents/Resources/config |



What seems to be happening is that the `../` as part of `__dirname` contributes to the resolved path (from `Contents/Resources/app`)... 

So: `Contents/Resources/app/../server/config/../../config`
Resolves to: `Contents/Resources/config`.

I had thought that perhaps the structure of `desktop/build/` could be rearranged in a way that mirrored the new relationship - move `desktop.js` (the built file) to `app/desktop/desktop.js` and unzip out the `app/calypso/` contents in to the root level. I hoped that this would allow the paths to resolve as they would in development but this didn't seem to work - `cwd()` was still `app` and so paths were still broken.

I feel like convenience took care of this in `wp-desktop` - the issue simply never arose because there was never a `../` in the `__dirname`.

I'm going to explore a few more options but it's quite a trial and error effort at the moment - I've struggled to find any material on this issue so far.

I'd love to hear of any thoughts from anyone who's seen similar issues before :)

spen commented:
I've made some progress with the latest commit - `make run` & `make osx` run nicely without hiccups.
I've added the blocked tag to this PR though, since neither `make test` or `make test-osx` as a result of the changes made in my latest commit. I'll explain why I think this is after explaining the last batch of changes.

> I had thought that perhaps the structure of desktop/build/ could be rearranged in a way that mirrored the new relationship - move desktop.js (the built file) to app/desktop/desktop.js and unzip out the app/calypso/ contents in to the root level. I hoped that this would allow the paths to resolve as they would in development but this didn't seem to work - cwd() was still app and so paths were still broken.

Towards the end of https://github.com/Automattic/wp-calypso/pull/12430#issuecomment-292670362 I mentioned that calling `process.cwd()` would always result in `Contents/Resources/app`.

I traced the reason for this - We were forcing a move (`chdir`) to this path in [`desktop/desktop/env.js`](https://github.com/Automattic/wp-calypso/pull/12430/files#diff-851aaafe3855ad3883a2020c0fd99318R32). As previously mentioned, resolving `path.resolve( __dirname, '..', '..', 'config' )` from here was resulting in a location that didn't exist and pointing outside of the `app/` directory.

By shuffling files around in `desktop/build` to better match the relationship between Calypso and desktop and changing directory to `app/desktop` instead of just `app` I found that these path resolving issues were... resolved!

Unfortunately this does disrupt the tests though. 
We see the following error message:

```bash
// This seems to be thrown from a chdir polyfill when the target location isn't found.
Error: ENOENT: no such file or directory, uv_chdir
```

Looking at the source of the error I see that it comes from the line that I changed in `desktop/desktop/env.js`. I beleive that this is due the `/desktop` directory not existing in the electron app that's used be `electron-mocha` as when I print out the `apppath` variable in this file it shows this location:

```bash
calypso-nested/desktop/node_modules/electron/dist/Electron.app/Contents/Resources
```

Here's the contents of that directory... note that there isn't an `app` directory because this is straight from the `node_modules` Electron.

<img width="1153" alt="screen shot 2017-04-10 at 03 00 27" src="https://cloud.githubusercontent.com/assets/4335450/24844000/9ada36d6-1d9f-11e7-951e-7b24b723d743.png">

I think I'm heading in the right direction by matching the structure of `build` with how Calypso is laid out but will have to have a closer look at the tests to resolve the problems that we're seeing here before this is ready to merge.

spen commented:
Edit: I've added a more comprehensive list of tasks to the opening description ☝

spen commented:
#### Update: 
In the latest commit - https://github.com/Automattic/wp-calypso/pull/12430/commits/c1b53e9e80b0c38e2411c407e8a94eb5bbcc44c8 - I allow for the `apppath` in `desktop/desktop/env.js` to be overridden, with what was previously `CALYPSO_PATH`. I've renamed that `DESKTOP_APP_PATH` to better reflect it's use.

It seems that `CALYPSO_PATH` was only ever used through the `make test` flow - during that flow `process.chdir` is called twice is quick succession - first in `desktop/desktop/env.js`, then in `desktop/desktop/test/boot.js`.

With the changes made in https://github.com/Automattic/wp-calypso/pull/12430/commits/d83daf82aa3d36f36e2ff47f1b0fc8671795f7df, the first `process.chdir` would error as `app/desktop` doesn't exist in the default app that `electron-mocha` uses.

With the latest changes, we avoid this by changing to `CALYPSO_PATH` (or rather `DESKTOP_APP_PATH`) in both places.

spen commented:
It's been a while since I've given an update directly here, so here are a few points around the tasks left to tackle in the final stretch:

### Package Linux

`make package-linux` is the last of what I've been calling the `main make commands` that still has work.
I've not been able to test this myself but @astralbodies has reported an error at runtime:
```
Error: Cannot find module 'uglify-js'
```
With the majority of the issues thus far relating to mismatching of file paths, the isn't the first time deps have been missing.
This does strike me as a little odd though, as I'm not sure what the purpose of 'uglify-js' is in this context... post packaging!

### Docs

Almost there! I've been working on the remaining doc changes on and off inbetween tackling some of the more technical problems that remain.
The chunkiest part of what is remaining is handling the `CREDITS.md` file (#13373). Though we could just keep both for the time being I started going down a little 🐰🕳 here and I'm now looking at first updating the formatting of Calypsos `CREDIT.md` and having this file be generated automatically. 
If this becomes a blocker I'd push for keeping both until that PR is resolved.

### Linting

My first inkling with fixing up desktops lint warnings was to simple work through the results of `make lint` but I soon realised there was a flaw. Simply running `make lint` (essentially `eslint ./desktop`) lets ESLint search for config files (`.eslintrc` etc.). When ESLint scans, it picks up both `/desktop/.eslintrc` and the root `.eslintrc.js`.
I next thought about amending the line in the Makefile to only use `/desktop/.eslintrc` but those rules have [hardly been updated in the past year](https://github.com/Automattic/wp-desktop/blame/master/.eslintrc) - compare this to the continuous efforts made in Calypso itself and it seems obvious that falling back to desktops original configuration isn't the way to go.

There may be a few caveats, for example `'use-strict';` was required in the past because of the version of node Electron ran with, but I think the goal should be to fully remove the second `.eslintrc`.

### CircleCI

I have to admit - I'm not very experienced with CircleCI. My first pass at integrating desktops circle config in to Calypsos (#13325) show some issues...
As it stands the build takes around 9 minutes making it the longest to run and pushing the total suite time up significantly.
I'm sure that there are ways to optimize so this time could be reduced but I would guess that desktop will always be the longest (and thus limiting) process, simply because there's more to bundle and package.
It could be that running this in the CI on every commit is not the solution - at least not until the desktop process itself is streamlined.

I'll be away for a few days (🏄!) but my priority on return will be squishing that `package-linux` command as I see this as the biggest blocker right now - the other remaining issues could really be tackled later at a push and should be much easier to review given that they won't require such extensive testing.

astralbodies commented:
@spen Do you know if you'll be able to get time to have this merged after you're out of the support rotation?

spen commented:
@astralbodies Hoping so!
Almost done there so I'll soon be speaking with @jblz and co about what happens next (We spoke briefly about this yesterday actually).

I think the biggest hurdle left to jump is testing on linux... which I'll need to figure out myself to be able to debug my way through the remaining OS related issues.
I've not gotten around to that just yet as I need to buy a cheapo linux machine to test on or get a VM running on my new MBP when it lands (I've had a bunch of problems getting one to run on my current machine).

Fingers are very much crossed that the rebase isn't too hellish :P

jblz commented:
>  linux machine to test on or get a VM running on my new MBP when it lands

I've had good success installing Ubuntu & Fedora on VirtualBox on the Mac.  Let me know if you hit a snag.

> Fingers are very much crossed that the rebase isn't too hellish :P

> Files changed 171

![homer-disappearing-into-a-hedge](https://user-images.githubusercontent.com/1587282/26952169-3fc5937e-4c72-11e7-9487-caa1fb51a52f.gif)

alisterscott commented:
Closing as this PR is stale - please reopen if necessary
